### DISCUSSION TITLE
Using the library when the JSON data can be successful or an error message

### DISCUSSION THREAD

**Discussion by ManjitBedi** [freshOS/Networking#61] [Q&A]
Hi,

I am currently writing some code where I am using the Networking library to fetch code from a server using an access token and other parameters.  If there is a problem with the body data in the request, an error response will be returned as JSON.

How do I could something using Networking & Combine to handle the data that comes back that can be one of two different responses per se.  I am using something like this at the moment:

```swift

    /// - Parameters: dictionary with access token etc
    func somePost(paramsIn: Params) -> AnyPublisher<Any, Error> {
        post("/endpoint", params: paramsIn)
    }

api.somePost(paramsIn: params).sink { result in
            switch result {
            case .finished:
                break
            case .failure(let err):
                completion(err)
            }
        } receiveValue: { [self] responseData in
            // check what the response data is.
            if let err = checkForErrorsInReponse(data: data as! [String : Any]) {
                completion(err)
            } else {
                // Check what's in the data
               processResponse((data: data as! [String : Any])
                completion(nil)
            }
        }
        .store(in: &cancellables)
```

This code is not going to work - well it requires a lot of boilderplate;  the data in the closure is a Dictionary not JSON data.  I did try changing the type of `responseData`  to Any but in the debugger it is indicating that it is still a dictionary. 

How. might I rewrite the code using a JSON publisher? 

```let jsonPublisher: AnyPublisher<Any, Error> = client.get("")```

**Comment by s4cha** [comment]
Hi ManjitBedi, please excuse the delay. From what I see you are using AnyPublisher<Any, Error> which is the "json" one already. 
As for handling errors, if the status code is not in the 200 range, the error is going to be in the `failure` block and not the `receiveValue`. You can have more info here https://github.com/freshOS/Networking#handling-errors.

In the failure block, in order to access the error JSON, you need to parse the `Error` into `NetworkingError`, then you will have access to  `status`, `code` & `jsonPayload`.

Here is an example using the async await api of parsing the error JSON Payload and bridging it to a Domain Error.

```swift
public func validate(user: User, with token: String) async throws {
    do {
        return try await put("/users/\(user.id)/validate", body: .urlEncoded(["token": token]))
    } catch {
        if let e = error as? NetworkingError, let payload = e.jsonPayload, let jsonData = try? JSONSerialization.data(withJSONObject: payload, options: []) {
            let decoder = JSODecoder()
            let validationError = try? decoder.decode(ValidationErrorResponseJSON.self, from: jsonData)
            if validationError?.error == "EmailAlreadyValidated" {
                throw MyDomainError.emailAlreadyValidated
            }
        }
        throw MyDomainError.unknown
    }
}
```

Hope it helps :)

**Reply by ManjitBedi** [reply]
Thanks for the reply.  Since then, I have seen elsewhere that you can set up an enum in a struct to then handle if the response from the server has an error like an expired token. 

To reiterate 

- a successful response is received but in the body of the response there is an error.
- the JSON decoding to a struct does not work currently as there can be two different types of payloads in the response for a success or server error of some kind specific to the back end

I don't know how I can code this up efficiently to handle the multiple payloads.  For the moment, I am get the data from the response then decoding the JSON etc.  It is creating more boiler plate code.
